Use timeout per rule, instead of per file

aryx opened the pull request:
test plan:
```
$ semgrep-core -timeout 2 -l python -rules
tests/OTHER/perf1/timeout_gracefully.yaml
tests/OTHER/perf1/timeout_gracefully.py
...
[1.363  Warning    Main.Memory_limit    ] large heap size: 431 MiB (memory limit is 0 MiB). If a crash follows, you could suspect OOM.
[2.005  Info       Main.Match_rules     ] timeout for rule forcetimeout on file tests/OTHER/perf1/timeout_gracefully.py
tests/OTHER/perf1/timeout_gracefully.py:26 with rule succeed
     return redirect(url)
tests/OTHER/perf1/timeout_gracefully.py:14 with rule succeed
     return redirect(request.get_host())
tests/OTHER/perf1/timeout_gracefully.py:39 with rule succeed
     return redirect(url)
tests/OTHER/perf1/timeout_gracefully.py:33 with rule succeed
     return redirect(url)
tests/OTHER/perf1/timeout_gracefully.py:19 with rule succeed
     return redirect(request.method)
tests/OTHER/perf1/timeout_gracefully.py:9 with rule succeed
     return redirect(request.POST.get("next"))
[2.006  Trace      Main.Run_semgrep     ] done with tests/OTHER/perf1/timeout_gracefully.py
[2.006  Info       Main.Run_semgrep     ] found 6 matches, 1 errors, 0 skipped targets
tests/OTHER/perf1/timeout_gracefully.py:1:0: Timeout:
```
where it used to just finish without any match and just a timeout error.


PR checklist:
- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

aryx committed:
Use timeout per rule, instead of per file

test plan:
```
$ semgrep-core -timeout 2 -l python -rules
tests/OTHER/perf1/timeout_gracefully.yaml
tests/OTHER/perf1/timeout_gracefully.py
...
[1.363  Warning    Main.Memory_limit    ] large heap size: 431 MiB (memory limit is 0 MiB). If a crash follows, you could suspect OOM.
[2.005  Info       Main.Match_rules     ] timeout for rule forcetimeout on file tests/OTHER/perf1/timeout_gracefully.py
tests/OTHER/perf1/timeout_gracefully.py:26 with rule succeed
     return redirect(url)
tests/OTHER/perf1/timeout_gracefully.py:14 with rule succeed
     return redirect(request.get_host())
tests/OTHER/perf1/timeout_gracefully.py:39 with rule succeed
     return redirect(url)
tests/OTHER/perf1/timeout_gracefully.py:33 with rule succeed
     return redirect(url)
tests/OTHER/perf1/timeout_gracefully.py:19 with rule succeed
     return redirect(request.method)
tests/OTHER/perf1/timeout_gracefully.py:9 with rule succeed
     return redirect(request.POST.get("next"))
[2.006  Trace      Main.Run_semgrep     ] done with tests/OTHER/perf1/timeout_gracefully.py
[2.006  Info       Main.Run_semgrep     ] found 6 matches, 1 errors, 0 skipped targets
tests/OTHER/perf1/timeout_gracefully.py:1:0: Timeout:
```
where it used to just finish without any match and just a timeout error.

aryx commented:
with semgrep: 
```
semgrep --timeout 2 --config timeout_gracefully.yaml timeout_gracefully.py 
Running 2 rules...
Warning: 1 timeout error(s) in timeout_gracefully.py when running the following rules: [forcetimeout]

 timeout_gracefully.py 
     succeed
        This rule will not time out


          9┆     return redirect(request.POST.get("next"))
          ⋮┆----------------------------------------
         14┆     return redirect(request.get_host())
          ⋮┆----------------------------------------
         19┆     return redirect(request.method)
          ⋮┆----------------------------------------
         26┆     return redirect(url)
          ⋮┆----------------------------------------
         33┆     return redirect(url)
          ⋮┆----------------------------------------
         39┆     return redirect(url)
ran 2 rules on 1 files: 6 findings
found problems analyzing 1 file; run with --verbose for details or run with --strict to exit non-zero if any file cannot
be analyzed cleanly

```

IagoAbal commented:
Nice! We should fix the changelog, right now it says that these timeouts won't work.

Need to remove this:

> Semgrep will not recover from timeouts in one rule by running others

aryx commented:
ah ok will do. There are lots of refactorings and cleanup to do actually. I just wanted to get the minimal PR first and check CI.

aryx committed:
update changelog

aryx committed:
start factorizing search_rules and tainting_rules

aryx committed:
Merge branch 'develop' into pad/pa-822-fix-timeout-errors

aryx committed:
Add -timeout_threshold to semgrep-core

This is needed now that we pass all the rules to semgrep-core.

test plan:
pytest tests/e2e/test_check.py::test_timeout_threshold  now works

aryx committed:
handle timeout_threshold = 0

aryx committed:
Make Match_tainting_rules.check_rule like Match_search_rules

This will allow to factorize the timeout management code.

aryx committed:
get rid of Rule_timeout exn

aryx committed:
Handle timeout for tainting rules too

aryx committed:
Merge branch 'develop' into pad/pa-822-fix-timeout-errors

aryx committed:
merge changelog
